test/e2e/clusterctl_upgrade.go

```diff
@@ -387,14 +390,48 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			input.PreWaitForCluster(managementClusterProxy, testNamespace.Name, workLoadClusterName)
 		}
 
+		machineCRD := &apiextensionsv1.CustomResourceDefinition{}
+		if err := managementClusterProxy.GetClient().Get(ctx, client.ObjectKey{Name: "machines.cluster.x-k8s.io"}, machineCRD); err != nil {
+			Expect(err).ToNot(HaveOccurred(), "failed to retrieve a machine CRD")
+		}
+
+		// Build GroupVersionKind for Machine resources
+		machineListGVK := schema.GroupVersionKind{
+			Group: machineCRD.Spec.Group,
+			Kind:  machineCRD.Spec.Names.ListKind,
+		}
+
+		// Sort the machineCRD.Spec.Versions slice
+		sort.Slice(machineCRD.Spec.Versions, func(i, j int) bool {
+			return machineCRD.Spec.Versions[i].Name < machineCRD.Spec.Versions[j].Name
+		})
+
+		// Pick the first served version
+		for _, gvk := range machineCRD.Spec.Versions {
+			if gvk.Served {
```

Comment:
Let's use the storage version instead as proposed by stefan

```suggestion
			if gvk.Storage {
```

Reply:
The advantage is that the storage version is slightly more future-proof as it would also work after the provider upgrade (a storage version cannot be directly dropped or unserved during an upgrade)

Reply:
Hmm, after my last change (sorting), `pull-cluster-api-e2e-workload-upgrade-1-27-latest-main` passed this time, maybe it is a luck and we saw https://github.com/kubernetes-sigs/cluster-api/pull/9097#issuecomment-1673026825